crates/sqllib/src/aggregates.rs

```diff
@@ -351,25 +351,141 @@ macro_rules! universal_aggregate {
 }
 pub(crate) use universal_aggregate;
 
+#[doc(hidden)]
+pub fn agg_min__<T>(left: T, right: T) -> T
+where
+    T: Ord + Clone + Debug,
+{
+    left.min(right)
+}
+
+universal_aggregate!(agg_min, T where T: Ord + Clone + Debug);
+
 #[doc(hidden)]
 pub fn agg_max__<T>(left: T, right: T) -> T
 where
-    T: Ord + Clone,
+    T: Ord + Clone + Debug,
 {
     left.max(right)
 }
 
-universal_aggregate!(agg_max);
+universal_aggregate!(agg_max, T where T: Ord + Clone + Debug);
+
+fn o0<L, R>(t: (L, R)) -> (Option<L>, R) {
+    (Some(t.0), t.1)
+}
+
+// Macro to create variants of an aggregation function
+// There must exist a function f__(left: (L, R), right: (L, T)) -> (L, R)
+// This creates 3 more functions
+// f_N_<L, R>(left: (<Option<L>, R), right: (L, R)) -> (Option<L>, R)
+// etc.
+// And 4 more functions:
+// f_N_N_conditional<L, R>(left: (L, R), right: (L, R), predicate: bool) -> (L, R)
+macro_rules! universal_aggregate2 {
+    ($func:ident, $l: ty, $r: ty where $($bounds:tt)*) => {
+        ::paste::paste! {
+            #[doc(hidden)]
+            pub fn [<$func _N_ >]<$l, $r>( left: (Option<$l>, $r), right: ($l, $r)) -> (Option<$l>, $r)
+                where $($bounds)*
+            {
+                match left {
+                    (None, _) => o0(right.clone()),
+                    (Some(left), r) => o0([<$func __>]((left, r), right)),
+                }
+            }
+
+            #[doc(hidden)]
+            pub fn [<$func __N>]<$l, $r>( left: ($l, $r), right: (Option<$l>, $r) ) -> (Option<$l>, $r)
+                where $($bounds)*
+            {
+                match right {
+                    (None, _) => o0(left.clone()),
+                    (Some(right), r) => o0([<$func __>](left, (right, r))),
+                }
+            }
+
+            #[doc(hidden)]
+            pub fn [<$func _N_N>]<$l, $r>( left: (Option<$l>, $r), right: (Option<$l>, $r) ) -> (Option<$l>, $r)
+                where $($bounds)*
+            {
+                match (left.clone(), right.clone()) {
+                    ((None, _), _) => right.clone(),
+                    (_, (None, _)) => left.clone(),
+                    ((Some(left), l1), (Some(right), r1)) => o0([<$func __>]((left, l1), (right, r1))),
+                }
```

Comment:
This isn't an unnecessary number of clones, at least? I'd think that one could at least do:
```suggestion
                match (left.clone(), right.clone()) {
                    ((None, _), right) => right,
                    (left, (None, _)) => left,
                    ((Some(left), l1), (Some(right), r1)) => o0([<$func __>]((left, l1), (right, r1))),
                }
```